plugin/src/plugin.rs

```diff
@@ -0,0 +1,173 @@
+    ) -> &mut Self {
+        self.option.insert(RpcOption {
+            name: name.to_string(),
+            opt_typ: opt_type.to_string(),
+            default: def_val,
+            description: description.to_string(),
+            deprecated,
+        });
+        self
+    }
+
+    // FIXME: adding the long description as parameter
+    pub fn add_rpc_method<F: 'static>(
+        &'a mut self,
+        name: &str,
+        usage: &str,
+        description: &str,
+        callback: F,
+    ) -> &mut Self
+    where
+        F: RPCMethod<T> + 'static,
+    {
+        self.rpc_method.insert(name.to_owned(), Box::new(callback));
+        self.rpc_info.insert(RPCMethodInfo {
+            name: name.to_string(),
+            usage: usage.to_string(),
+            description: description.to_string(),
+            long_description: description.to_string(),
+            deprecated: false,
+        });
+        self
+    }
+
+    fn call_rpc_method(&'a mut self, name: &str, params: &serde_json::Value) -> serde_json::Value {
+        let command = self.rpc_method.get(name).unwrap().clone();
+        command.call(self, params)
+    }
+
+    pub fn register_hook<F: 'static>(
+        &'a mut self,
+        hook_name: &str,
+        before: Option<Vec<String>>,
+        after: Option<Vec<String>>,
+        callback: F,
+    ) -> &mut Self
+    where
+        F: RPCMethod<T> + 'static,
+    {
+        self.rpc_hook
+            .insert(hook_name.to_owned(), Box::new(callback));
+        self.hook_info.insert(RPCHookInfo {
+            name: hook_name.to_owned(),
+            before,
+            after,
+        });
+        self
+    }
+
+    pub fn register_notification<F: 'static>(&mut self, name: &str, callback: F) -> &mut Self
+    where
+        F: 'static + RPCMethod<T> + Clone,
+    {
+        self.rpc_nofitication
+            .insert(name.to_owned(), Box::new(callback));
+        self
+    }
+
+    pub fn start(&'a mut self) {
+        let reader = io::stdin();
+        let mut writer = io::stdout();
+        let mut buffer = String::new();
+
+        self.rpc_method
+            .insert("getmanifest".to_owned(), Box::new(ManifestRPC {}));
+        self.rpc_method
+            .insert("init".to_owned(), Box::new(InitRPC {}));
+        // FIXME: core lightning end with the double endline, so this can cause
+        // problem for some input reader.
+        // we need to parse the writer, and avoid this while loop
+        loop {
+            let _ = reader.read_line(&mut buffer);
+            let req_str = buffer.to_string();
+            if req_str.trim().is_empty() {
+                continue;
+            }
+            buffer.clear();
+            let request: Request<serde_json::Value> = serde_json::from_str(&req_str).unwrap();
+            let response = self.call_rpc_method(request.method, &request.params);
+            let mut rpc_response = init_success_response(request.id);
+            rpc_response["result"] = response;
```

Comment:
There is no way for the user right now to return an error, and I think will be a good API, to give the user the possibility to return an Option, so in this way, the user can return an error at any point! with simply add `Err(...)`

But maybe this is something that I need to track with a separate issue